src/librustc/hir/map/definitions.rs

```diff
@@ -91,6 +92,10 @@ impl DefPathTable {
 pub struct Definitions {
     table: DefPathTable,
     node_to_def_index: NodeMap<DefIndex>,
+    pub hir_to_def_index: HirIdMap<DefIndex>,
+    /// `DefIndex`es created by DefCollector::create_def; used only to complete the hir_to_def_index
+    /// mapping after the lowering
+    defs_awaiting_hir_id: NodeMap<DefIndex>,
```

Comment:
This is another reason why I would prefer if everything with a `DefId` was a `HirId` root: you wouldn't need any mappings, a HIR node with a `DefId` would have a `HirId` of `(def_id, 0)`.
cc @michaelwoerister @nikomatsakis @petrochenkov

Reply:
IIRC, we wanted `HirId` owners to correspond to item-likes mainly because that's the incremental compilation granularity we were shooting for.

Reply:
I see, but that seems like it's introducing artificial complexity (with perhaps the exception of `TypeckTables` - but even in that case, we could just have separate `TypeckTables` for closures, that are kept in a map in the parent body's `TypeckTables`).

Bonus: right now things like embedded constants have their own `TypeckTables` but share a `HirId` space with their enclosing item, meaning their intra-item IDs are not in a nice `0..n` range (if they were, we could maybe use something more efficient than hashmaps).